### ISSUE TITLE
Flickering

### ISSUE THREAD

**Issue by alexeckermann** [alexeckermann/esp32-lamp#1]
During some testing scenarios there is visible flickering on the LED strip. [See video demonstrating flickering.](https://raw.githubusercontent.com/alexeckermann/esp32-lamp/main/other/issue-1/active-wifi.mov)

Testing was based on the [test/neopixel-test](https://github.com/alexeckermann/esp32-lamp/tree/main/test/neopixel-test) directory.

### Software

There was some early permutations of the flickering that was due to software issues.

Early code was too slow which meant one tick from the hardware timer would begin before the previous one had finished writing to GPIO. Slowing the tick cycle and adjusting the tick handler to be more performant have removed this particular issue. This was confirmed by reviewing start and end time of the tick function via serial printing.

**NeoPixel library in MicroPython**

Whilst debugging this issue I have had a look at the [underlying NeoPixel library in MicroPython](https://github.com/micropython/micropython/blob/master/ports/esp32/espneopixel.c). From my initial impressions, the generation of the WS281X data is very time dependent. It looks as though there are some measures to request a stable environment for timing sensitive features but I am not sure wether this is respected.

An aside here is that the LED strip I am using is WS2811 requiring the 400kHz data frequency. By default the library uses 800kHz which definitely causes some flickering (among other unexpected behaviour).

**Wireless modem activity**

There is a clear correlation between the WiFi modem being active and the flickering. [Here is a video showing the same animation when WiFi is not active](https://github.com/alexeckermann/esp32-lamp/blob/main/other/issue-1/no-wifi.mov?raw=true).

There have been some occasions where there has been a very infrequent flicker when WiFi was not active. Anecdotally, it was one flicker per 60 seconds as opposed to one every 5 seconds when WiFi was active in testing.

I ran a test where I sent ICMP packets to the ESP32 when it was on the network to see if that caused more flickering. Being anecdotal I cant say one way or another, it was very subtle and I am unsure if there was any difference at all. The ping responses were very slow (>100ms) for a device on the same network and varied quite a bit (30–300ms).

### Hardware

First point of call was all physical connections. The LED strip is connected via JST connectors which I crimped and installed. There is a chance that there is some less than ideal connections. A simple jiggle test on the cables didn't yield any obvious changes in flickering behaviour or frequency.

I am unsure wether there is any issues with the logic level converter part of the board. Is the 5V supply too high? Is the 330ohm resistor wrong? Do I need some kind of filter cap? No idea on any of these counts.

Another thought was wether the GPIO pin (12) was somehow now appropriate for this use. After consulting the ESP32 Thing pinout I tried a couple of sibling pins but saw no meaningful improvement. This could still be an option to test more pins.

## Theories

These are my theories in order of what I think is most likely:

1. The timing sensitive GPIO operations are being interrupted by other tasks on the CPU related to wireless IO
2. Something I don't know about the logic level converter is corrupting the data
3. Theres a connection issue with the JST connectors or elsewhere in the board
4. Using a hardware timer or my current software implementation is causing an unexpected interruption to GPIO operations
5. There maybe better output on a different GPIO pin
6. RF interference on the data line

I'd appreciate any insights into what I could be experiencing or if there are any tests I can perform to narrow the possible causes.

**Comment by alexeckermann** [comment]
It might be important to note that I don't have a logic analyser. To properly debug this issue I feel I may have to start observing the data on the line to determine if I have a data timing issue.

Any recommendations on a simple and relatively affordable analyser would be greatly appreciated.

**Comment by alexeckermann** [comment]
_(I'm posting these comments 'in the blind' to help clarify my own thoughts and processes on this project. It may help someone in future. But for now its helping me keep track of this weekend project.)_

After going through some research on the topic it looks as though that this is a common issue and it points mostly to two things:

1. The logic level converter used currently in this project may have some impact on the pulses.
    - [Testing level shifters for running Adafruit NeoPixel strips from a 3.3 Volt Teensy](https://happyinmotion.com/?p=1247)

2. The current implementation of the pulse generator in the NeoPixel module on MicroPython may not be operating within the specification under certain conditions (CPU load, IO).
    - Various issues on the micropython project around NeoPixels. Most would like to use FastLED and/or RMT for pulse generation.

Switching the logic level converter is an easy and low cost change. I have chosen to use the `74AHCT125` as specified in the [Adafruit NeoPixel Uberguide](https://learn.adafruit.com/adafruit-neopixel-uberguide/best-practices). This should rule out the external hardware side of this issue by using a proven and recommended part. A good plus learning that not all logic level converters are made equal. I have a simple logic analyser coming along with this purchase which may help observe the changes and identify the issue.

The second issue will be somewhat more time intensive to resolve. From the outset, since I am very new to this level of software I would ideally like to avoid writing/building firmware unless its the only path forward. With that said, this issue points to a solution based in utilising the RMT feature of the ESP32. There are some examples elsewhere but nothing rock solid enough to be a reference library/module.

In conclusion, with the new logic level converter installed the remaining exploration is to write a RMT based NeoPixel driver. Initially I will focus on using the existing MicroPython RMT class and see how far I can get with that. If it's not usable at that high level I may have to go down to writing a C module and embedding it in a custom firmware build. I'd like to avoid the latter as it will likely take some time for me to work out.

**Comment by alexeckermann** [comment]
Good progress in the past week.

This flickering issue has been resolved :tada: The primary cause was the provided NeoPixel library in MicroPython can't generate pulse timing accurately.

The solution was to write an implementation of a NeoPixel driver using RMT. I was initially concerned that this would require a C module. But, the provided RMT class in Python has and should be performant enough for any purpose.

Thanks to @nevercast (et al.) [for this gist](https://gist.github.com/nevercast/9c48505cc6c5687af59bcb4a22062795) that was the basis of my initial tests and was a great example for writing my own implementation.

I did purchase a new logic level converter but since the RMT implementation has resolved the issue I will hold off on the change as it may not be necessary.

To close this issue I would like to finish a base class for writing to WS281(1|2) lights that anyone can use. That should come together over the coming week.
